Add generic fast format converters for better performance

taldcroft opened the pull request:
## Description

Creating a `CxoTime` object is unfortunately slow, taking a few hundred microsecs. Often the only reason for this it so convert to a different format like `date` or `secs`. This PR includes functionality to do those conversions in tens of microsecs (for the scalar case).

Features:
* Includes a functional interface `convert_time_format` that converts between `date`, `secs`, `jd`, `maude` and `greta` formats without creating a `CxoTime` object. It can also infer the input time format in a performant way for common formats (in particular `date` and `secs`).
* Generalize the convenience methods `date2secs` and `secs2date` fast converters to include all combinations of the supported fast formats in `convert_time_format`.
* Update the `CxoTime` initializers so that `CxoTime(tm)` returns the `tm` object if it is already a `CxoTime` instance. This allows writing code like `start = CxoTime(start)` with no performance penalty if `start` is already a `CxoTime`.  `CxoTime(tm, copy=True)` will make a copy as expected.'

This is based on the `remove-cxotime-fast-parser` branch since that was needed. The intent is to rebase on master once #34  is merged.

### Todo
- [x] Docs
- [x] Add unit tests
- [x] Ensure reasonable docstrings for the convenience conversion functions
- [x] Some code clean-up

## Interface impacts
<!-- API changes, file format updates, coordination of changes with the community. -->
The `date2secs` and `secs2date` functions were removed from the `cxotime.cxotime` so `from cxotime.cxotime import date2secs` will now fail. This was not the public interface and code should be using `from cxotime import date2secs`.

## Testing
<!-- If relevant describe any special setup for testing. -->

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
- [x] Mac


Independent check of unit tests by Jean
- [x] Linux (in a dev ska3-matlab-2023.4rc6 env)

### Functional tests
<!-- Describe and document results of any functional tests, otherwise leave the text below -->
Roughly speaking using `CxoTime` for a single value is around 350 µs, while the "fast" functions will take around 25 µs.
```
>>> from cxotime import *
>>> %timeit CxoTime(100.0).greta
373 µs ± 6.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> %timeit secs2greta(100.0)
24.4 µs ± 577 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> %timeit convert_time_format(100.0, 'greta')
28.1 µs ± 1.71 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> %timeit convert_time_format(100.0, 'greta', fmt_in='secs')
24.2 µs ± 457 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> %timeit greta2secs(2020001.123456789)
25.4 µs ± 1.85 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

>>> %timeit greta2secs('2020001.123456789')
20.7 µs ± 1.61 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Ultimate conversion speed is about 0.2 µs / conversion for large arrays:
```
>>> x = ["2020001.123456789"] * 1000
>>> %timeit greta2secs(x)
358 µs ± 5.43 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

>>> x = np.array(["2020001.123456789"] * 1000)
>>> %timeit greta2secs(x)
218 µs ± 4.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

jeanconn commented on cxotime/convert.py:
```diff
@@ -0,0 +1,343 @@
+
+
+def get_format(val, fmt_out):
+    fmt_in = None
```
Right, similarly in https://github.com/sot/cxotime/pull/34 I had trouble figuring out how much I was supposed to know about astropy time internals.  Maybe cxotime should have some more doc links to astropy time docs?  For jd1 and jd2 probably this helps: https://docs.astropy.org/en/stable/time/index.html#internal-representation

taldcroft replied:
Fixed with a more detailed docstring.

javierggt commented on cxotime/convert.py:
```diff
@@ -0,0 +1,326 @@
+                x[:4] + x[5:8] + "." + x[9:11] + x[12:14] + x[15:17] + x[18:21]
+                for x in date.flat
+            ]
+        )
```
```
dtf2d = np.frompyfunc(single_dtf2d, 7, 1)

fun = np.frompyfunc(lambda x: f"{x[:4]}{x[5:8]}.{x[9:11]}{x[12:14]}{x[15:17]}{x[18:21]}", 1, 1)


def convert_jd1_jd2_to_greta_2(jd1, jd2):
    date = convert_jd1_jd2_to_date_2(jd1, jd2)
    # Convert '1997:365:23:58:57.816' to '1997365.235857816'
    #          012345678901234567890
    if isinstance(date, np.ndarray):
        out = fun(date)
        out.shape = date.shape
    else:
        x = date  # 15 ns
        out = x[:4] + x[5:8] + "." + x[9:11] + x[12:14] + x[15:17] + x[18:21]  # 660 ns
    return out


def convert_jd1_jd2_to_date(jd1, jd2):
    iys, ims, ids, ihmsfs = erfa.d2dtf(b"TT", 3, jd1, jd2)
    ihrs = ihmsfs["h"]
    imins = ihmsfs["m"]
    isecs = ihmsfs["s"]
    ifracs = ihmsfs["f"]

    if isinstance(jd1, np.ndarray):
        dates = dtf2d(iys, ims, ids, ihrs, imins, isecs, ifracs)
        out = np.array(dates).reshape(jd1.shape)
    else:
        yday = datetime.datetime(iys, ims, ids).timetuple().tm_yday
        out = f"{iys:4d}:{yday:03d}:{ihrs:02d}:{imins:02d}:{isecs:02d}.{ifracs:03d}"

    return out

```

javierggt commented on docs/index.rst:
```diff
@@ -125,39 +100,141 @@ argument.
+
+::
+
+    $ python
```
if you remove this, the code is rendered as python code, with proper colors, and the user can copy/paste.

javierggt commented on docs/index.rst:
```diff
@@ -125,39 +100,141 @@ argument.
+  >>> convert_time_format("2022:001:00:00:00.123", "greta")
+  '2022001.000000123'
+  >>> convert_time_format(100.123, "date")
+ '1998:001:00:00:36.939'
```
I think this line screws up the alignment.

I think it is better to have four spaces in all anyway, like the code below.

jeanconn commented on cxotime/cxotime.py:
```diff
@@ -25,148 +24,6 @@
 iers.conf.auto_download = False
 
 
-def date2secs(date):
```
Is the issue that cxotime.cxotime.date2secs won't work unless there's an import here from convert but cxotime.date2secs will still work?

taldcroft replied:
I think all these `<fmt1>2<fmt2>` converters are available in `cxotime.cxotime` and in the top-level `cxotime` package namespace. I didn't look at the code again but from this branch it seems to work:
```
(ska3) ➜  cxotime git:(performance) ipython
Python 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:27:35) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from cxotime import date2secs
In [2]: from cxotime.convert import date2secs
In [3]: from cxotime.convert import date2greta

In [4]:
```

taldcroft replied:
Right. I think that `cxotime.cxotime.date2secs` was not technically the public API. `cxotime.date2secs` is the public interface and that still works.

(And sorry about the other comment you might have seen (now deleted) that was completely wrong).

jeanconn replied:
And this seems fine, I just wanted to make sure I was on the same page.

taldcroft commented:
I rebased this and tests are passing locally on Mac. All good? We'll see in integration testing if something comes up?

I updated the Interface section to note the removal of `cxotime.cxotime.date2secs/secs2date`.

jeanconn commented:
I'll do an independent unit test run with the latest to wrap up the testing requirements in the description.
